src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,471 @@
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            with opentelemetry_tracer.start_as_current_span(generate_span_id()) as span:
+                span_id = span.get_span_context().span_id
+                if span.parent:
+                    span_parent_id = span.parent.span_id
+                else:
+                    span_parent_id = None
+                parent_trace_metadata = TRACE_FLOW_CONTEXT.get(span_parent_id)
+                if parent_trace_metadata:
+                    TRACE_FLOW_CONTEXT[span_id] = FlowContext(
+                        span_id=span_id,
+                        trace_parent_id=span_parent_id,
+                        is_flow_log=False,
+                    )
+
+                # Write the Tool Kernel to the Span on HL_FILE_OT_KEY
+                span.set_attribute(HUMANLOOP_PATH_KEY, path if path else func.__name__)
+                span.set_attribute(HUMANLOOP_FILE_TYPE_KEY, "tool")
+                if tool_kernel:
+                    write_to_opentelemetry_span(
+                        span=span,
+                        key=f"{HUMANLOOP_FILE_KEY}.tool",
+                        value=tool_kernel,
+                    )
+
+                # Call the decorated function
+                try:
+                    output = func(*args, **kwargs)
+                    error = None
+                except Exception as e:
+                    logger.error(f"Error calling {func.__name__}: {e}")
+                    output = None
+                    error = str(e)
+
+                # Populate known Tool Log attributes
+                tool_log = {
+                    "inputs": args_to_inputs(func, args, kwargs),
+                    "output": output,
+                    "error": error,
+                }
+
+                # Write the Tool Log to the Span on HL_LOG_OT_KEY
+                write_to_opentelemetry_span(
+                    span=span,
+                    key=HUMANLOOP_LOG_KEY,
+                    value=tool_log,
+                )
+
+                # Return the output of the decorated function
+                return output
+
+        wrapper.file = File(  # type: ignore
+            path=path if path else func.__name__,
+            type="tool",
+            version=tool_kernel,
+            callable=wrapper,
+        )
+
+        return wrapper
+
+    return decorator
+
+
+def _build_tool_kernel(
+    func: Callable,
+    attributes: Optional[dict[str, Optional[Any]]],
+    setup_values: Optional[dict[str, Optional[Any]]],
+    strict: bool,
+) -> ToolKernelRequestParams:
+    """Build ToolKernelRequest object from decorated function."""
+    source_code = textwrap.dedent(inspect.getsource(func))
+    # Remove decorator from source code by finding first 'def'
+    # This makes the source_code extraction idempotent whether
+    # the decorator is applied directly or used as a higher-order
+    # function
+    source_code = source_code[source_code.find("def") :]
```

Comment:
i think this is brittle and we should wrap in a try-except.

E.g. this throws an error.

<img width="645" alt="image" src="https://github.com/user-attachments/assets/34812c59-8237-4e00-b6c5-746880a263c0">

<img width="744" alt="image" src="https://github.com/user-attachments/assets/a9616ef2-ff02-4302-979f-2c12e35c21f2">

while uncommon, i don't think it's that rare (as in building up functions with partial and then applying decorators manually)

Reply:
i'd rather raise with custom message warning partials shouldn't be used

Reply:
update: fixed per previous comment